### PR TITLE
EQ4-11 En tant qu'étudiant, je veux m'inscrire 

### PR DESCRIPTION
**Pull request by wjeam**
**Cette branche contient:**

- Le backend permettant d'enregistrer les données des étudiants dans la base de données
- Le frontend contenant un dialog permettant aux étudiants d'entrer les informations requis pour s'enregistrer sur la platforme

### CODE REVIEW

File: backend/src/main/java/com/team4/backend/service/UserService.java [hunk 1]
```diff
@@ -30,9 +31,14 @@ public UserService(UserRepository userRepository, PBKDF2Encoder pbkdf2Encoder, J
         return userRepository.findByEmailAndPasswordAndIsEnabledTrue(authRequestDto.getEmail(), pbkdf2Encoder.encode(authRequestDto.getPassword()))
                 .map(jwtUtil::generateToken)
                 .switchIfEmpty(Mono.error(new ResponseStatusException(HttpStatus.UNAUTHORIZED, "Can't find user with this credentials")));
+    }
 
+    public Mono<User> findByEmail(String email) {
```
Comment (0xD34DC0DE) [inline backend/src/main/java/com/team4/backend/service/UserService.java:36]: do we need this function (can we use the isEmailTaken method instead)

File: backend/src/main/java/com/team4/backend/service/UserService.java [hunk 2]
```diff
@@ -30,9 +31,14 @@ public UserService(UserRepository userRepository, PBKDF2Encoder pbkdf2Encoder, J
...
+        return userRepository.findByEmail(email);
     }
 
+    public Mono<Boolean> isEmailTaken(String email) { return userRepository.existsByEmail(email); }
```
Comment (0xD34DC0DE) [inline backend/src/main/java/com/team4/backend/service/UserService.java:40]: maybe rename the method to existsByEmail

File: backend/src/test/java/com/team4/backend/controller/StudentControllerTest.java [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    @Autowired
+    private WebTestClient webTestClient;
+
+    @MockBean
+    StudentService studentService;
+
+    @MockBean
+    UserService userService;
+
+
```
Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/controller/StudentControllerTest.java:37]: remove extra line

File: backend/src/test/java/com/team4/backend/controller/StudentControllerTest.java [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                );
+
+        when(userService.findByEmail(any(String.class))).thenReturn(Mono.empty());
+
+        webTestClient
+                .post().uri("/student/register").bodyValue(studentDto)
+                .exchange()
+                .expectStatus().isCreated()
+                .expectBody().isEmpty();
+
```
Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/controller/StudentControllerTest.java:63]: remove extra line

Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/controller/StudentControllerTest.java:63]: Add //ACT, //ASSERT comments

File: backend/src/test/java/com/team4/backend/controller/StudentControllerTest.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        when(userService.findByEmail(any(String.class))).thenReturn(Mono.just(alreadyExistingStudent));
+
+        webTestClient
+                .post().uri("/student/register").bodyValue(studentDto)
+                .exchange()
+                .expectStatus().isEqualTo(HttpStatus.CONFLICT)
+                .expectBody().isEmpty();
+
+    }
```
Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/controller/StudentControllerTest.java:83]: add AAA comments

File: backend/src/test/java/com/team4/backend/controller/UserControllerTest.java [hunk 6]
```diff
@@ -30,14 +29,14 @@
     UserService userService;
 
     @Test
-    public void login(){
+    public void login() {
         //ARRANGE
         AuthRequestDto authRequestDto = new AuthRequestDto("12345678@gmail.com", "massou123");
 
         when(userService.login(authRequestDto)).thenReturn(Mono.just("sadasdadsas"));
```
Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/controller/UserControllerTest.java:36]: use better testing values

File: backend/src/test/java/com/team4/backend/controller/UserControllerTest.java [hunk 7]
```diff
@@ -47,7 +46,26 @@ public void login(){
...
-        assertEquals(HttpStatus.OK,httpStatus);
+        assertEquals(HttpStatus.OK, httpStatus);
     }
 
+    @Test
+    public void isEmailTaken() {
+        // ARRANGE
+        String email = "123456789@gmail.com";
+
+        // ACT
```
Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/controller/UserControllerTest.java:57]: mocking is still part of //ARRANGE move comment down one line

File: backend/src/test/java/com/team4/backend/repository/UserRepositoryTest.java [hunk 8]
```diff
@@ -84,4 +85,24 @@ void findByEmailAndPasswordAndIsEnabledTrue() {
         StepVerifier.create(userMono2).expectNextCount(0).verifyComplete();
         StepVerifier.create(userMono3).expectNextCount(0).verifyComplete();
     }
+
+    @Test
+    void existsByEmail() {
+        // ARRANGE
+        String email1 = "123456789@gmail.com";
+        String email2 = "asdf@gmail.com";
```
Comment (0xD34DC0DE) [inline backend/src/test/java/com/team4/backend/repository/UserRepositoryTest.java:93]: Use better testing values

Comment (0xD34DC0DE) [reply backend/src/test/java/com/team4/backend/repository/UserRepositoryTest.java:93]: and reformat file

File: frontend/src/components/EmailFormField.js [hunk 9]
```diff
@@ -0,0 +1,62 @@
+import React, { useState, useEffect } from "react";
+import TextFormField from "./TextFormField";
+import axios from "axios";
+
+const emailRegexValidation =
+  /^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
```
Comment (0xD34DC0DE) [inline frontend/src/components/EmailFormField.js:6]: 😢

File: frontend/src/components/PasswordFormField.js [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+      setErrorMessage((errors) => ({ ...errors, confirmPassword: "" }));
+      valid(false);
+    } else {
+      setErrorMessage((errors) => ({
+        ...errors,
+        confirmPassword: "Les mots de passe ne sont pas identique",
+      }));
+      valid(false);
+    }
+  }, [field.password, field.confirmPassword, step, valid]);
```
Comment (0xD34DC0DE) [inline frontend/src/components/PasswordFormField.js:32]: I'm a bit confused here, why are the error messages going inside the "confirmPassword" variable, the variable name is a bit confusing

File: frontend/src/components/PasswordFormField.js [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+    }));
+    onFieldChange(event);
+  };
+
+  return (
+    <>
+      <TextFormField
+        focus={true}
+        id="password"
+        dialogContentText="Entrez votre mot de passe et confirmez-le"
```
Comment (0xD34DC0DE) [inline frontend/src/components/PasswordFormField.js:47]: suggestion: remove "et confirmez-le"

File: frontend/src/components/Register.js [hunk 12]
```diff
@@ -1,179 +1,144 @@
...
+      url: "http://localhost:8080/student/register",
+      data: {
+        email: form.email,
+        password: form.password,
+        phoneNumber: form.phoneNumber,
+        firstName: form.firstName,
+        registrationNumber: form.registrationNumber,
+        lastName: form.lastName,
+      },
+      responseType: "json",
```
Comment (0xD34DC0DE) [inline frontend/src/components/Register.js:60]: application/json ?

**Review by 0xD34DC0DE — CHANGES_REQUESTED** [review]

File: frontend/src/components/Register.js (hunk 12 above)
Comment (wjeam) [reply frontend/src/components/Register.js:60]: image

File: frontend/src/components/PasswordFormField.js (hunk 10 above)
Comment (wjeam) [reply frontend/src/components/PasswordFormField.js:32]: Error messages goes into confirmPassword as confirming a password implies that the "confirmPassword" field matches the "password" field, therefore the error is highlighted on the "confirmPassword" field.

Comment (0xD34DC0DE) [reply frontend/src/components/PasswordFormField.js:32]: so you are replacing the content of the field with the error message ?

File: backend/src/main/java/com/team4/backend/service/UserService.java (hunk 2 above)
Comment (wjeam) [reply backend/src/main/java/com/team4/backend/service/UserService.java:40]: image

File: backend/src/main/java/com/team4/backend/service/UserService.java (hunk 1 above)
Comment (wjeam) [reply backend/src/main/java/com/team4/backend/service/UserService.java:36]: image

File: backend/src/main/java/com/team4/backend/repository/UserRepository.java [hunk 13]
```diff
@@ -5,10 +5,13 @@
...
-
 @Repository
 public interface UserRepository extends ReactiveMongoRepository<User, String> {
     Mono<User> findByRegistrationNumberAndPassword(String registrationNumber, String password);
+
     Mono<User> findByEmailAndPasswordAndIsEnabledTrue(String email, String password);
+
+    Mono<User> findByEmail(String email);
+
+    Mono<Boolean> existsByEmail(String email);
```
[the remark this answers is not in the document]
Comment (wjeam) [reply backend/src/main/java/com/team4/backend/repository/UserRepository.java:16]: Both

File: backend/src/main/java/com/team4/backend/service/UserService.java (hunk 1 above)
Comment (0xD34DC0DE) [reply backend/src/main/java/com/team4/backend/service/UserService.java:36]: Was more of a personal note

File: frontend/src/components/PasswordFormField.js (hunk 10 above)
Comment (wjeam) [reply frontend/src/components/PasswordFormField.js:32]: No, replacing the error message underneath the field.

File: frontend/src/components/EmailFormField.js (hunk 9 above)
Comment (wjeam) [reply frontend/src/components/EmailFormField.js:6]: 😈

File: frontend/src/components/PasswordFormField.js (hunk 10 above)
Comment (0xD34DC0DE) [reply frontend/src/components/PasswordFormField.js:32]: so the variable has a special meaning in the context of setErrorMessage() ?

**Review by 0xD34DC0DE — APPROVED** [review]

**Pull request merged**